### PR TITLE
Add support for JRuby

### LINKED ISSUE #10: JRuby support
JRuby needs a -java platform version of io-console.

We have partial support for the io/console library currently, but not in gem form. Moving our code to a gem would not be difficult. We would maintain the code in this repository, but pushes of new releases would need to happen for both standard and -java gems (pushing only the standard one would break JRuby).

The JRuby code is pure-Ruby, using FFI where necessary: https://github.com/jruby/jruby/tree/master/lib/ruby/stdlib/io

### PR DESCRIPTION
**Pull request by headius**
This PR contains the JRuby implementation of io/console based on FFI. It should be pushed as a `java` platform gem any time the io-console gem is updated.

Fixes #10.

### CODE REVIEW

**Comment by headius** [comment]
One bit of process I'm not sure about: this -java gem can only be built with JRuby, and likewise the standard gem can't be build on JRuby, because the gemspec uses RUBY_ENGINE. This could be modified to allow building both versions on any impl, since they're just packaging files (i.e. there's no compile step required for either JRuby or MRI).

**Comment by nobu** [comment]
Do you want io-console to include the Java version?
Otherwise, it would be possible to install a stub file for JRuby, I think, like https://github.com/ruby/io-console/tree/jruby-support

**Comment by headius** [comment]
@nobu I think it would be best for the gem to actually contain our version of the library, as we already do for psych, json, racc, and other libraries. We will have to push a `-java` platform gem either way, and moving the sources from JRuby into the gem means we can maintain it alongside the CRuby version.

We want to match CRuby's use of default/preinstalled gems as closely as possible going forward.

**Comment by headius** [comment]
Your addition of --platform will allow any implementation to build either platform gem, so I like that part.

File: lib/io/console/jruby/native_console.rb [hunk 1]
```diff
@@ -0,0 +1,153 @@
+# Load appropriate native bits for BSD or Linux
+case RbConfig::CONFIG['host_os'].downcase
+when /darwin|openbsd|freebsd|netbsd/
+  require 'io/console/bsd_console'
```
Comment (nobu) [inline lib/io/console/jruby/native_console.rb]: Isn't it 'io/console/jruby/bsd_console'?

Comment (headius) [reply lib/io/console/jruby/native_console.rb]: You are correct...I didn't realize these paths were used in the specific impls too. I will correct and re-push.

**Comment by headius** [comment]
I have re-pushed the primary commit with paths fixed. This version is now also live in JRuby proper and passing CI.

I also pushed a second commit that fixes a broken test on JRuby. The path to the "rubybin" executable was tweaked to substitute `/ruby/` for the name of the TTY-less executable, which caused the `jruby` executable to become `jjruby`. The test works as before on MRI.

**Comment by eregon** [comment]
I'd like to have this FFI implementation work on TruffleRuby too, see https://github.com/ruby/reline/issues/124 and https://github.com/oracle/truffleruby/issues/1904#issuecomment-585147711.

I think using the `java` platform wouldn't work for that, but simply use the default `ruby` platform and skip compilation on JRuby/TruffleRuby in extconf.rb.

File: lib/io/console/jruby/native_console.rb [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    end
+  end
+  private :ttymode_yield
+
+  TTY_RAW = Proc.new do |t|
+    LibC.cfmakeraw(t)
+    t[:c_lflag] &= ~(LibC::ECHOE|LibC::ECHOK)
+  end
+
+  def raw(*, &block)
```
Comment (nobu) [inline lib/io/console/jruby/native_console.rb]: The options are not supported.

**Comment by nobu** [comment]
Isn't it better to place these files under a different name, e.g. "ffilib/", if truffleruby will use them too?

**Comment by nobu** [comment]
I meant "not only for jruby".

**Comment by nobu** [comment]
I'm thinking of installing/moving ffi-version files by extconf.rb.

**Comment by eregon** [comment]
@nobu I think there should be a single gem push for simplicity, so a single platform.
I'd suggest having the files under `lib/io/console/ffi`, and always ship them.

**Comment by nobu** [comment]
I pushed 0.5.6, which just skips the extension on other than CRuby.
Merger of ffi version will be after it catches 0.5 up.

**Comment by eregon** [comment]
@nobu Thanks, I can confirm `gem install reline` works on TruffleRuby now.

**Comment by aycabta** [comment]
> I can confirm gem install reline works on TruffleRuby now.

Great! I guess that it's possible to install on JRuby too.

And, I think the `intr: ` option should be supported by the io-console gem of other platforms because it's used when users press Ctrl-C to reset an inputting text.

**Comment by headius** [comment]
So the issue with reline was resolved. The remaining issue would be io-console getting installed by a JRuby user, possibly through a transitive dependency. I would still like to see this gem contain the FFI implementation, so it could be shared across non-CRuby implementations. The FFI would be the fallback for platforms or runtimes that do not build the C extension. Some day, perhaps it could replace the C extension?

**Comment by eregon** [comment]
I agree it would be good to have the FFI implementation here too.

I think @nobu had some concern in https://github.com/ruby/io-console/pull/11#discussion_r379968629 and @aycabta in https://github.com/ruby/io-console/pull/11#issuecomment-588040025. @headius Could you try to address them so it's fully compatible with the C extension?

**Comment by headius** [comment]
I should be clear: this code was not originally written by me. I am hoping through this PR to release it into the public trust, so we can collaborate on making it a drop-in replacement for io/console.

I would welcome suggestions and patches for concrete improvements to make.

@nobu What do you mean "the options are not supported?" Are you referring to the LibC flags you linked? Or something like the `intr:` option @aycabta mentioned?

@aycabta Yes I agree this should match the C extension as closely as possible, but I'm not familiar with the `intr:` option.

File: lib/io/console/jruby/native_console.rb (hunk 2 above)
Comment (nobu) [reply lib/io/console/jruby/native_console.rb:46]: I mean `min`, `time` and `intr` options which are supported by `IO#raw`.

**Comment by headius** [comment]
I updated my branch to include the changes @nobu made as well as one fix from @aardvark179. Not much has changed since the original discussion of this PR (e.g. the `intr:` option is still not supported.

**Comment by headius** [comment]
In the short term, can we get a -java gem release that includes the FFI logic but not the C extension?

At the moment we can't install reline on Windows because of the io-console requirement, which forces users to have a `make` command installed even though nothing will be built on JRuby.

Our users have been ok with this version for many years, even if it does not have all features and does not pass all tests. We will take responsibility for the code and bug reports against the FFI logic (which will probably be filed with JRuby rather than io-console).

**Review by nobu — APPROVED** [review]

**Comment by headius** [comment]
For others following along...

I have implemented the missing functionality in the FFI io/console and it now runs green on CRuby. It also runs green on JRuby with the exception of a few tests that require PTY.spawn to produce a child process with its own TTY, which is broken on JRuby (https://github.com/jruby/jruby/issues/6552).

The FFI implementation can be shipped now, I believe.

**Pull request merged**